addons/l10n_ar_base/models/res_partner.py

```diff
@@ -0,0 +1,126 @@
+from odoo.tools.safe_eval import safe_eval
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    l10n_ar_cuit = fields.Char(
+        compute='_compute_l10n_ar_cuit',
+    )
+    l10n_ar_formated_cuit = fields.Char(
+        compute='_compute_l10n_ar_formated_cuit',
+    )
+    l10n_ar_id_number = fields.Char(
+        string='Identification Number',
+    )
+    l10n_ar_id_category_id = fields.Many2one(
+        string="Identification Category",
+        comodel_name='l10n_ar_id_category',
+        index=True,
+        auto_join=True,
+    )
+
+    @api.multi
+    def cuit_required(self):
+        """ Return the cuit number is this one is defined if not raise an
+        UserError
+        """
+        self.ensure_one()
+        if not self.l10n_ar_cuit:
+            raise UserError(_(
+                'No CUIT configured for partner [%i] %s') % (
+                    self.id, self.name))
+        return self.l10n_ar_cuit
+
+    @api.depends('l10n_ar_cuit')
+    def _compute_l10n_ar_formated_cuit(self):
+        """ This will add some dash to the CUIT number in order to show in his
+        natural format: {person_category}-{number}-{validation_number}
+        """
+        for rec in self:
+            if not rec.l10n_ar_cuit:
+                continue
+            cuit = rec.l10n_ar_cuit
+            rec.l10n_ar_formated_cuit = "{0}-{1}-{2}".format(
+                cuit[0:2], cuit[2:10], cuit[10:])
+
+    @api.depends('l10n_ar_id_number', 'l10n_ar_id_category_id')
+    def _compute_l10n_ar_cuit(self):
+        """ We add this computed field that returns cuit or nothing ig this one
+        is not set for the partner. This validation can be also dony by calling
+        cuit_required() method that returns the cuit nombre of error if this
+        one is not found.
+        """
+        for rec in self:
+            # If the partner is outside Argentina then we return the defined
+            # country cuit defined by AFIP for that specific partner
+            if rec.l10n_ar_id_category_id.afip_code != 80:
+                country = rec.country_id
+                if country and country.code != 'AR':
+                    if rec.is_company:
+                        rec.l10n_ar_cuit = country.l10n_ar_cuit_juridica
+                    else:
+                        rec.l10n_ar_cuit = country.l10n_ar_cuit_fisica
+                continue
+            if rec.l10n_ar_id_category_id.afip_code == 80:
+                rec.l10n_ar_cuit = rec.l10n_ar_id_number
+
+    @api.constrains('l10n_ar_id_number', 'l10n_ar_id_category_id')
+    def check_vat(self):
+        """ Update the the vat field using the information we have from
+        l10n_ar_id_number and l10n_ar_id_category_id fields
+        """
+        for rec in self:
+            if rec.l10n_ar_id_number and rec.l10n_ar_id_category_id and \
+               rec.l10n_ar_id_category_id.afip_code == 80:
+                rec.vat = 'AR' + rec.l10n_ar_id_number
```

Comment:
setting a field in a constraint?

Reply:
Till [this PR](https://github.com/odoo/odoo/pull/31059) is accepted we are using this hack to set up fields avoiding to inherit create and write methods. If you have a better suggestion they are welcome.
We also do this in account_document so the decision will guide as in many cases.

Reply:
ok, we are going to change to create/write methods